datafusion/core/src/physical_optimizer/pruning.rs

```diff
@@ -903,6 +1004,42 @@ impl<'a> PruningExpressionBuilder<'a> {
         self.required_columns
             .max_column_expr(&self.column, &self.column_expr, self.field)
     }
+
+    /// Note that this function intentionally overwrites the column expression to [`phys_expr::Column`].
+    /// i.e. expressions like [`phys_expr::CastExpr`] or [`phys_expr::TryCastExpr`] will be overwritten.
+    ///
+    /// This is to avoid cases like `cast(x_null_count)` or `try_cast(x_null_count)`.
```

Comment:
If the query predicate is `WHERE cast(x AS data_type)`, the original rewrite will turn `x_min` into `cast(x_min)` and `x_max` into `cast(x_max AS data_type)`. Same for `try_cast()`. See here:

https://github.com/appletreeisyellow/arrow-datafusion/blob/chunchun/pruning-predicate-column-known-tobe-null/datafusion/core/src/physical_optimizer/pruning.rs#L1075-L1101

We don't want `cast()` and `try_cast()` to rewrite `x_null_count` into `cast(x_null_count AS data_type)`, so this step is to avoid `cast(x_null_count AS data_type)` from happening

Reply:
Hmm, I don't quite get this part too. Why we need to handle `cast` or `try_cast` here specially? If the predicate is `cast(x as dt)`, shouldn't we keep it as is?

Reply:
I checked how this function is used below. I think it is not used to rewrite predicate expression. It is to actually simply retune the null_count physical expression no matter the predicate expression is.

Reply:
I feel that the current comment is confused. It is better to rephrase it a bit.